# Profiling resnets on GPU - Training and Data Loading

Next, we compare the performance of ResNets with VGG16.

In [ ]:
import numpy as np
import torch
from torch import nn
from torch.profiler import profile, record_function, ProfilerActivity
import torchvision
import torchvision.transforms as transforms
import torch.optim as optim
import time

In [ ]:
path = "./data"
workers = 2
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
batch_size = 64

In [ ]:
transform_train = transforms.Compose(
    [transforms.ToTensor(),
     transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))])
trainset = torchvision.datasets.CIFAR100(
    root=path, train=True, download=True, transform=transform_train)
trainloader = torch.utils.data.DataLoader(
    trainset, batch_size=batch_size, shuffle=True, num_workers=workers)

testset = torchvision.datasets.CIFAR100(
    root=path, train=False, download=True, transform=transform_train)
testloader = torch.utils.data.DataLoader(
    testset, batch_size = batch_size, shuffle=True, num_workers=workers)

Files already downloaded and verified


Files already downloaded and verified


In [ ]:
def train(model, criterion, trainloader, optimizer, scheduler, epochs):
  model.train()
  for epoch in range(0, epochs):
      print('\nEpoch: %d' % epoch)
      running_loss = 0.0

      for i, data in enumerate(trainloader, 0):
          inputs, labels = data
          inputs = inputs.to(device)
          labels = labels.to(device)

          optimizer.zero_grad()

          outputs = model(inputs)
          loss = criterion(outputs,labels)
          loss.backward()
          optimizer.step()

          running_loss += loss.item()

          del inputs, labels

          if i % 100 == 99:
              print(f'Average loss: {running_loss / 100}')
              running_loss = 0.0

      scheduler.step()

In [ ]:
def evaluate(model, testloader):
    model.eval()
    total = 0
    correct = 0
    
    with torch.no_grad():
        for data in testloader:
            images, labels = data
            images = images.to(device)
            labels = labels.to(device)
            # calculate outputs by running images through the network
            outputs = model(images)
            # the class with the highest energy is what we choose as prediction
            _, predicted = torch.max(outputs.data, 1)
            total += labels.size(0)
            correct += (predicted == labels).sum().item()

    print(f'Accuracy of the network on the test images: {100 * correct // total} %')

In [ ]:
lr = 0.0001
epochs = 30

net = torchvision.models.resnet101()
net = net.to(device)

criterion = nn.CrossEntropyLoss()
optimizer = optim.Adam(net.parameters(), lr=lr)
scheduler = torch.optim.lr_scheduler.MultiStepLR(optimizer, milestones=[20], gamma=0.1)

# Train
print("Training Large ResNet-50 with bottleneck layers")
start = time.perf_counter()
train(net, criterion, trainloader, optimizer, scheduler, epochs)
stop = time.perf_counter()

# Evaluate
evaluate(net, testloader)

print(f'Total training time: {(stop-start)/60} minutes')

Training Large ResNet-50 with bottleneck layers

Epoch: 0


Average loss: 5.060655860900879


Average loss: 4.725819172859192


Average loss: 4.6744813013076785


Average loss: 4.62241265296936


Average loss: 4.547175297737121


Average loss: 4.456430416107178


Average loss: 4.383152432441712



Epoch: 1


Average loss: 4.258324599266052


Average loss: 4.214882514476776


Average loss: 4.184802825450897


Average loss: 4.116050727367401


Average loss: 4.102784156799316


Average loss: 4.087313041687012


Average loss: 4.048307416439056



Epoch: 2


Average loss: 3.9540627646446227


Average loss: 3.9242401456832887


Average loss: 3.89469122171402


Average loss: 3.8767638206481934


Average loss: 3.917403779029846


Average loss: 3.853049817085266


Average loss: 3.8487058997154238



Epoch: 3


Average loss: 3.7145815753936766


Average loss: 3.6960347843170167


Average loss: 3.664842164516449


Average loss: 3.6488359117507936


Average loss: 3.6528082299232483


Average loss: 3.6568361210823057


Average loss: 3.6082517457008363



Epoch: 4


Average loss: 3.463297152519226


Average loss: 3.458758828639984


Average loss: 3.453449523448944


Average loss: 3.4817060303688048


Average loss: 3.455802581310272


Average loss: 3.4638257598876954


Average loss: 3.391943416595459



Epoch: 5


Average loss: 3.2625087571144102


Average loss: 3.2918419432640076


Average loss: 3.284442889690399


Average loss: 3.283922245502472


Average loss: 3.269281783103943


Average loss: 3.264394600391388


Average loss: 3.2745434761047365



Epoch: 6


Average loss: 3.0211448311805724


Average loss: 3.0397919702529905


Average loss: 3.0861644387245177


Average loss: 3.0973438525199892


Average loss: 3.1091872358322146


Average loss: 3.1060520029067993


Average loss: 3.0815226793289185



Epoch: 7


Average loss: 2.820487480163574


Average loss: 2.857542922496796


Average loss: 2.912879045009613


Average loss: 2.9208518195152284


Average loss: 2.97535352230072


Average loss: 2.9001335215568544


Average loss: 2.9116913533210753



Epoch: 8


Average loss: 2.6495044767856597


Average loss: 2.6948070120811463


Average loss: 2.7030054831504824


Average loss: 2.742064185142517


Average loss: 2.712270402908325


Average loss: 2.7502348804473877


Average loss: 2.750917077064514



Epoch: 9


Average loss: 2.4615309584140777


Average loss: 2.4345881080627443


Average loss: 2.486353998184204


Average loss: 2.518462882041931


Average loss: 2.5203849864006043


Average loss: 2.5672487008571623


Average loss: 2.566840032339096



Epoch: 10


Average loss: 2.1897380650043488


Average loss: 2.264831329584122


Average loss: 2.2989568090438843


Average loss: 2.301047749519348


Average loss: 2.330163117647171


Average loss: 2.349996410608292


Average loss: 2.3936985778808593



Epoch: 11


Average loss: 1.9451926910877229


Average loss: 2.031562342643738


Average loss: 2.0775553727149965


Average loss: 2.1231157898902895


Average loss: 2.148930997848511


Average loss: 2.168696655035019


Average loss: 2.1367018866539



Epoch: 12


Average loss: 1.749840428829193


Average loss: 1.768986897468567


Average loss: 1.8680131709575654


Average loss: 1.8737841868400573


Average loss: 1.8952578938007354


Average loss: 1.9589737927913666


Average loss: 1.9383213996887207



Epoch: 13


Average loss: 1.4818607032299043


Average loss: 1.5572206246852875


Average loss: 1.583930168747902


Average loss: 1.6179425263404845


Average loss: 1.6485005629062652


Average loss: 1.6691915416717529


Average loss: 1.7018457686901092



Epoch: 14


Average loss: 1.2878988963365554


Average loss: 1.2854948794841767


Average loss: 1.3881088709831237


Average loss: 1.434444043636322


Average loss: 1.4686547696590424


Average loss: 1.5205434262752533


Average loss: 1.4654507553577423



Epoch: 15


Average loss: 1.048441436290741


Average loss: 1.093756291270256


Average loss: 1.1504930847883224


Average loss: 1.1590350210666656


Average loss: 1.2133644276857376


Average loss: 1.321761916279793


Average loss: 1.312170587182045



Epoch: 16


Average loss: 0.870326144695282


Average loss: 0.9011977547407151


Average loss: 0.9583460354804992


Average loss: 0.9744516870379448


Average loss: 1.0645183199644088


Average loss: 1.0806925481557845


Average loss: 1.1279363238811493



Epoch: 17


Average loss: 0.7490478411316872


Average loss: 0.8045013463497162


Average loss: 0.8462964046001434


Average loss: 0.8691300559043884


Average loss: 0.920678145289421


Average loss: 0.8852011120319366


Average loss: 0.9166160264611244



Epoch: 18


Average loss: 0.6248911571502686


Average loss: 0.664103738963604


Average loss: 0.6813170865178109


Average loss: 0.6820065677165985


Average loss: 0.7353479158878327


Average loss: 0.7568830224871635


Average loss: 0.800679063796997



Epoch: 19


Average loss: 0.5495472621917724


Average loss: 0.5429289355874062


Average loss: 0.6050364819169044


Average loss: 0.6301738399267197


Average loss: 0.6613091164827347


Average loss: 0.6700724205374717


Average loss: 0.7180824798345565



Epoch: 20


Average loss: 0.40804426416754724


Average loss: 0.341467762440443


Average loss: 0.3012010062485933


Average loss: 0.2740486289560795


Average loss: 0.2760455070436001


Average loss: 0.2552720572799444


Average loss: 0.25558829851448533



Epoch: 21


Average loss: 0.17928668305277826


Average loss: 0.17423190847039222


Average loss: 0.1730066703632474


Average loss: 0.17350855357944966


Average loss: 0.16795401502400636


Average loss: 0.1682005316391587


Average loss: 0.16490217473357915



Epoch: 22


Average loss: 0.14939593199640513


Average loss: 0.12902913816273212


Average loss: 0.12280679267831146


Average loss: 0.12728378772735596


Average loss: 0.13506162786856293


Average loss: 0.12600708670914174


Average loss: 0.1338738962262869



Epoch: 23


Average loss: 0.10101016838103533


Average loss: 0.11718228248879313


Average loss: 0.10546889925375581


Average loss: 0.09438653126358985


Average loss: 0.10881256207823753


Average loss: 0.10233314061537385


Average loss: 0.1060936451330781



Epoch: 24


Average loss: 0.07968985889106989


Average loss: 0.08767623148858547


Average loss: 0.09595038261264563


Average loss: 0.08743364041671157


Average loss: 0.0877315555792302


Average loss: 0.08451070135459304


Average loss: 0.09639681714586913



Epoch: 25


Average loss: 0.07816302524879574


Average loss: 0.07359813610091806


Average loss: 0.08171036878600717


Average loss: 0.07797404620796441


Average loss: 0.08151495524682105


Average loss: 0.07702049132436514


Average loss: 0.08095038217492402



Epoch: 26


Average loss: 0.08192586204968393


Average loss: 0.07761169290170074


Average loss: 0.07969222642015666


Average loss: 0.07448916266672313


Average loss: 0.08429568842984736


Average loss: 0.0746065605385229


Average loss: 0.07220680152066052



Epoch: 27


Average loss: 0.07221817022189497


Average loss: 0.06987017534673214


Average loss: 0.067098378688097


Average loss: 0.05494449283927679


Average loss: 0.061003631916828456


Average loss: 0.06767832888290286


Average loss: 0.057660024333745244



Epoch: 28


Average loss: 0.05681521773338318


Average loss: 0.06975233418866993


Average loss: 0.04370523457415402


Average loss: 0.0556111620599404


Average loss: 0.04887798510491848


Average loss: 0.055429852064698935


Average loss: 0.05853318592533469



Epoch: 29


Average loss: 0.05217147413641214


Average loss: 0.05010019244160503


Average loss: 0.05537704872433096


Average loss: 0.05452247702050954


Average loss: 0.05160826716106385


Average loss: 0.05909354745410383


Average loss: 0.05428020820487291


Accuracy of the network on the test images: 30 %
Total training time: 23.239619165066628 minutes


In [ ]:
lr = 0.0001
epochs = 30

net = torchvision.models.vgg16()
net = net.to(device)

criterion = nn.CrossEntropyLoss()
optimizer = optim.SGD(net.parameters(), lr=lr, momentum=0.9)
scheduler = torch.optim.lr_scheduler.MultiStepLR(optimizer, milestones=[20], gamma=0.1)

# Train
print("Training VGG-16")
start = time.perf_counter()
train(net, criterion, trainloader, optimizer, scheduler, epochs)
stop = time.perf_counter()

# Evaluate
evaluate(net, testloader)

print(f'Total training time: {(stop-start)/60} minutes')

Training VGG-16

Epoch: 0


Average loss: 6.906110858917236


Average loss: 6.902322573661804


Average loss: 6.898639583587647


Average loss: 6.894145908355713


Average loss: 6.888122639656067


Average loss: 6.881031723022461


Average loss: 6.865200352668762



Epoch: 1


Average loss: 6.737410154342651


Average loss: 5.849116821289062


Average loss: 5.204124474525452


Average loss: 5.066347455978393


Average loss: 4.974654283523559


Average loss: 4.949827733039856


Average loss: 4.912020192146302



Epoch: 2


Average loss: 4.864743165969848


Average loss: 4.868194088935852


Average loss: 4.82703709602356


Average loss: 4.820792388916016


Average loss: 4.8111689043045045


Average loss: 4.797281875610351


Average loss: 4.788915419578553



Epoch: 3


Average loss: 4.775445327758789


Average loss: 4.782615146636963


Average loss: 4.779842185974121


Average loss: 4.773708143234253


Average loss: 4.769507617950439


Average loss: 4.760654339790344


Average loss: 4.748029956817627



Epoch: 4


Average loss: 4.7505575466156005


Average loss: 4.737827463150024


Average loss: 4.731775932312011


Average loss: 4.747884082794189


Average loss: 4.743182611465454


Average loss: 4.728683843612671


Average loss: 4.73349057674408



Epoch: 5


Average loss: 4.719668321609497


Average loss: 4.721165518760682


Average loss: 4.726508293151856


Average loss: 4.717904162406922


Average loss: 4.714796252250672


Average loss: 4.716151823997498


Average loss: 4.712210087776184



Epoch: 6


Average loss: 4.704454216957092


Average loss: 4.705355520248413


Average loss: 4.703031840324402


Average loss: 4.698498454093933


Average loss: 4.7094663858413695


Average loss: 4.70639018535614


Average loss: 4.6976695728302005



Epoch: 7


Average loss: 4.69336877822876


Average loss: 4.699126811027527


Average loss: 4.693798890113831


Average loss: 4.702457733154297


Average loss: 4.688880047798157


Average loss: 4.691206479072571


Average loss: 4.695767126083374



Epoch: 8


Average loss: 4.682473349571228


Average loss: 4.690956196784973


Average loss: 4.687723875045776


Average loss: 4.676143403053284


Average loss: 4.6869216680526735


Average loss: 4.68059280872345


Average loss: 4.682726287841797



Epoch: 9


Average loss: 4.684394283294678


Average loss: 4.683434600830078


Average loss: 4.680408821105957


Average loss: 4.678564848899842


Average loss: 4.677202425003052


Average loss: 4.673506150245666


Average loss: 4.674706168174744



Epoch: 10


Average loss: 4.674747834205627


Average loss: 4.678997101783753


Average loss: 4.681338758468628


Average loss: 4.670258383750916


Average loss: 4.674246587753296


Average loss: 4.673577399253845


Average loss: 4.674278950691223



Epoch: 11


Average loss: 4.67289698600769


Average loss: 4.671770825386047


Average loss: 4.667781844139099


Average loss: 4.671381978988648


Average loss: 4.663965058326721


Average loss: 4.6656540536880495


Average loss: 4.660555815696716



Epoch: 12


Average loss: 4.65875883102417


Average loss: 4.66448037147522


Average loss: 4.662559657096863


Average loss: 4.660728859901428


Average loss: 4.656481680870056


Average loss: 4.655019674301148


Average loss: 4.659320583343506



Epoch: 13


Average loss: 4.647920451164246


Average loss: 4.658445372581482


Average loss: 4.649120674133301


Average loss: 4.6553773021698


Average loss: 4.649468417167664


Average loss: 4.651583023071289


Average loss: 4.653269248008728



Epoch: 14


Average loss: 4.639553046226501


Average loss: 4.645528826713562


Average loss: 4.635546259880066


Average loss: 4.623284449577332


Average loss: 4.622096567153931


Average loss: 4.619272303581238


Average loss: 4.615611400604248



Epoch: 15


Average loss: 4.586213698387146


Average loss: 4.584472632408142


Average loss: 4.560019788742065


Average loss: 4.556256484985352


Average loss: 4.534865312576294


Average loss: 4.524986462593079


Average loss: 4.509026608467102



Epoch: 16


Average loss: 4.499568853378296


Average loss: 4.482790222167969


Average loss: 4.4888162469863895


Average loss: 4.477498126029968


Average loss: 4.4672762966156006


Average loss: 4.455148434638977


Average loss: 4.456248803138733



Epoch: 17


Average loss: 4.446076135635376


Average loss: 4.424253888130188


Average loss: 4.424100909233093


Average loss: 4.432634191513062


Average loss: 4.409874291419983


Average loss: 4.418293504714966


Average loss: 4.417320380210876



Epoch: 18


Average loss: 4.387074255943299


Average loss: 4.392570486068726


Average loss: 4.392613439559937


Average loss: 4.378586955070496


Average loss: 4.3831260919570925


Average loss: 4.363671913146972


Average loss: 4.381879954338074



Epoch: 19


Average loss: 4.352216062545776


Average loss: 4.342442755699158


Average loss: 4.326251215934754


Average loss: 4.327895503044129


Average loss: 4.33731297492981


Average loss: 4.314210081100464


Average loss: 4.314272413253784



Epoch: 20


Average loss: 4.271084218025208


Average loss: 4.266603059768677


Average loss: 4.277462611198425


Average loss: 4.259915490150451


Average loss: 4.256468286514282


Average loss: 4.280327739715577


Average loss: 4.270378022193909



Epoch: 21


Average loss: 4.26405731678009


Average loss: 4.254919638633728


Average loss: 4.2586262011528015


Average loss: 4.253804321289063


Average loss: 4.264138650894165


Average loss: 4.241846089363098


Average loss: 4.260636038780213



Epoch: 22


Average loss: 4.2420172595977785


Average loss: 4.228988316059112


Average loss: 4.250825939178466


Average loss: 4.259174127578735


Average loss: 4.242787098884582


Average loss: 4.241021451950073


Average loss: 4.231883928775788



Epoch: 23


Average loss: 4.21734215259552


Average loss: 4.225740382671356


Average loss: 4.240274891853333


Average loss: 4.231585144996643


Average loss: 4.2409640049934385


Average loss: 4.228636116981506


Average loss: 4.2439282560348515



Epoch: 24


Average loss: 4.22042407989502


Average loss: 4.220138857364654


Average loss: 4.222665891647339


Average loss: 4.22522186756134


Average loss: 4.224782817363739


Average loss: 4.221660504341125


Average loss: 4.211849200725555



Epoch: 25


Average loss: 4.204224984645844


Average loss: 4.245470416545868


Average loss: 4.214180507659912


Average loss: 4.191026179790497


Average loss: 4.222207112312317


Average loss: 4.205748255252838


Average loss: 4.198640546798706



Epoch: 26


Average loss: 4.216622896194458


Average loss: 4.189723870754242


Average loss: 4.199317493438721


Average loss: 4.187373898029327


Average loss: 4.195090494155884


Average loss: 4.196574945449829


Average loss: 4.207670848369599



Epoch: 27


Average loss: 4.203052251338959


Average loss: 4.204566330909729


Average loss: 4.18589834690094


Average loss: 4.189097933769226


Average loss: 4.186044239997864


Average loss: 4.17350697517395


Average loss: 4.191311690807343



Epoch: 28


Average loss: 4.180405528545379


Average loss: 4.181842224597931


Average loss: 4.190861754417419


Average loss: 4.195587961673737


Average loss: 4.173498368263244


Average loss: 4.166017138957978


Average loss: 4.187101199626922



Epoch: 29


Average loss: 4.175513677597046


Average loss: 4.178512210845947


Average loss: 4.181291816234588


Average loss: 4.170658078193664


Average loss: 4.170826919078827


Average loss: 4.15742039680481


Average loss: 4.176987178325653


Accuracy of the network on the test images: 6 %
Total training time: 9.758820507849972 minutes


## Analysis

When training the VGG16 model and the ResNet101 models, our final result is that ResNet101 took about 2.5 times longer to train, but achieved a significantly better accuracy. Testing accuracy after a very shallow training period with minimal hyperparameter tuning resulted in 30% accuracy, while the VGG achieved only 6%. The ResNet is a substantially larger model and much deeper, and is thus able to learn much better at the cost of total training time. This is a tradeoff between efficiency and performance.